### DISCUSSION TITLE
browser mutes sound on a unsecured http connection

### DISCUSSION THREAD

**Discussion by mithrendal** [vAmigaWeb/vAmigaWeb#255] [Q&A]
hi @wepl just discovered when launching vAmigaWeb on whdload.de on an unsecured connection (i.e. http) then the browser denies rendering the audio on a separate thread. 

to get browser play the audio vAmigaWeb has to be embedded by a page loaded from `https` protocol.

unsecure - no audio - browser denies playing audio
http://www.whdload.de

secured - audio works
https://www.whdload.de

in case somebody got on your site by `http` then you could simply redirect the `http` request with http 301 response code to `https` on your web server to fix this.

**Comment by wepl** [comment]
Thanks for the info, good to know.
But I don't like to redirect to https. People should use http if they want, e.g. on old machines.

**Reply by mithrendal** [reply]
I see. I forgot about that. Yeah leave it as it is ...

**Comment by mithrendal** [comment]
On click on the player which starts the emulator we could let the player recognize that it is loaded from unsecure and bring a message dialog box 

“Sound is only played by the browser when loaded from a secure https connection. Would you like to reload this page from secure connection ?”
`Yes` `No`

It is easy to bake this into the player … What do you think ?

**Reply by wepl** [reply]
this sounds great, would improve usebility I think 😄

**Reply by mithrendal** [reply]
Cool I baked it into the player …


Looks like this now 
<img width="1366" alt="image" src="https://github.com/user-attachments/assets/2467659c-4e64-4ffe-b0d9-02c88068b88c">

Is the text okay ?

**Reply by wepl** [reply]
yes, it's fine

**Comment by mithrendal** [comment]
```javascript
        if(window.location.protocol === 'http:')
        {
            if (window.confirm(
`Sound will only play if this page is loaded over a secure HTTPS connection.
\nWould you like me to reload the page using a secure connection now ?`
            )) 
            {
                window.location.replace('https://' + window.location.host + window.location.pathname + window.location.search);
            }
        }

```

this is what I added.

>People should use http if they want, e.g. on old machines.

we could also do an auto upgrade to https when we detect a modern browser. In that case there would be no message box. We would simply check the existence of some modern web APIs like `audioWorklet`. When existent and protocol was http then do automatically a redirect when the page is initially loaded.

**Comment by mithrendal** [comment]
vAmigaWeb connects to the modern AudioWorkletAPI which is supported since 2021 in all major browsers. 

See here for reference 

https://caniuse.com/?search=audioworklet


you could place the following snippet on the whdload html page 


```html
<script>
if (
  typeof AudioContext !== 'undefined' && 
  'audioWorklet' in AudioContext.prototype && 
  window.location.protocol === 'http:'
) 
{
    window.location.replace('https://' + window.location.host + window.location.pathname + window.location.search);
}
</script>
```


it checks for the modern API in the used browser. browsers before 2018 would always skip the https upgrade. All browsers after 2021 would immediately upgrade to https without any user interaction.  😎

**Comment by wepl** [comment]
Thanks, I have added that snippet.
I'm using chrome which always redirects to https himself. So no real check by me.

**Reply by mithrendal** [reply]
>I'm using chrome which always redirects to https himself. So no real check by me.

in chrome you have to type explicitly http:// before the URL then you will get the unsafe connection. You also will get that connection if some other sites out there link explicitly to http:// 

I have tested and script does not work 🫢

that is because the 2. condition is false in case of http

```javascript 
'audioWorklet' in AudioContext.prototype
```

the browser developers were clever and totally removed the audioWorklet API when using http rather then to block the usage.
 
So we can not really check if AudioAPI would be there on https while we are still on http . 

🤔 we could check the presence of WebAssembly instead, because that API is also supported on unsafe http. 

```javascript 
'WebAssembly' in window
```

Older browsers like `AWeb` which we like to stay on `http` do not support WebAssembly, so thats a sufficient condition to detect modern browsers.

lets try ... remove the check on `audioWorklet API `above and replace it with check on  `WebAssembly API `

**Reply by wepl** [reply]
ok, I changed it that way, works in Safari now 😄
